### PR TITLE
module: Patch imports when modules are loaded later

### PR DESCRIPTION
**Pull request by Zangetsu38**
- module: Patch imports when modules are loaded later

### CODE REVIEW

File: vita3k/interface.cpp [hunk 1]
```diff
@@ -272,6 +273,22 @@ static ExitCode load_app_impl(Ptr<const void> &entry_point, HostState &host, Gui
 
     host.renderer->features.hardware_flip = host.cfg.hardware_flip;
 
+    // Load user driver
```
Comment (frangarcj) [inline vita3k/interface.cpp:276]: Use an array similar to pre-loaded libraries in order to add more.

extract reading file and loading to a common function to this and pre-loaded

Comment (Zangetsu38) [reply vita3k/interface.cpp:276]: is not possible, need use diferent things, no same partition/folder need obligatory using totaly diferent code

**Review by frangarcj — DISMISSED** [review]

File: vita3k/interface.cpp [hunk 2]
```diff
@@ -272,27 +273,75 @@ static ExitCode load_app_impl(Ptr<const void> &entry_point, HostState &host, Gui
...
             } else
                 return FileNotFound;
         } else {
-            LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
+            LOG_DEBUG("User Pre-load module at \"{}\" not present", module_path);
+        }
+    }
+
+    // Load pre-loaded libraries
+    const auto module_app_path{ fs::path(host.pref_path) / "ux0/app" / host.io.title_id / "sce_module" };
```
Comment (sunho) [inline vita3k/interface.cpp:299]: Why not just create a temp variable and std::string(temp) + module_path to eliminate those duplicated code.

Comment (Zangetsu38) [reply vita3k/interface.cpp:299]: sorry, no understund what you want ask 🤔

**Comment by sunho** [comment]
The commit also moves the code from user to kernel so change the commit name lol.

**Comment by Zangetsu38** [comment]
> 
> 
> The commit also moves the code from user to kernel so change the commit name lol.

look more commit name, all is ewrited :)

File: vita3k/interface.cpp (hunk 2 above)
Comment (sunho) [reply vita3k/interface.cpp:299]:
```cpp

 if (fs::exists(module_app_path) && !fs::is_empty(module_app_path)) {
   device= VitaIoDevice::app0;
   prefix = std::string("app0:");
} else {
  device = VitaIoDevice::vs0;
  prefix = std::string("vs0:");
}
        const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;

            if (vfs::read_file(device, module_buffer, host.pref_path, host.io.title_id, module_path)) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];

                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }

```

Comment (sunho) [reply vita3k/interface.cpp:299]: just a pseudocode no able to use computer

Comment (Zangetsu38) [reply vita3k/interface.cpp:299]: look more, is not possible like that, path for module is diferent

Comment (sunho) [reply vita3k/interface.cpp:299]: then you can also set lib_load_list before

Comment (Zangetsu38) [reply vita3k/interface.cpp:299]: path module is diferent i said, i can't like that

Comment (sunho) [reply vita3k/interface.cpp:299]: I meant

```
if (fs::exists(module_app_path) && !fs::is_empty(module_app_path)) {
   device= VitaIoDevice::app0;
   prefix = std::string("app0:");
const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

} else {
  device = VitaIoDevice::vs0;
  prefix = std::string("vs0:");
 const char *const lib_load_list[] = {
            "sys/external/libc.suprx",
            "sys/external/libfios2.suprx",
            "sys/external/libult.suprx",
        };
}

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;

            if (vfs::read_file(device, module_buffer, host.pref_path, host.io.title_id, module_path)) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];

                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }
```

Comment (Zangetsu38) [reply vita3k/interface.cpp:299]: AHH yes, i have n,o thinks like that but your code no can works here, need inisalise it before if lol

Comment (sunho) [reply vita3k/interface.cpp:299]: you can make it work if you tweak it little, right? I believe in you. 😎

**Comment by sunho** [comment]
Oh it has it's hidden lol. Never mind.

File: vita3k/interface.cpp (hunk 2 above)
Comment (Zangetsu38) [reply vita3k/interface.cpp]: check again, i have reworks it, with diferent approch, and adios this char*...  i hate char xD

Comment (sunho) [reply vita3k/interface.cpp]: looks great to me. btw we've decided to not load driver_user by default.

Comment (sunho) [reply vita3k/interface.cpp]: I'm not sure whether it will cause regressions or not even. Can you test loading it with many games.

Comment (Zangetsu38) [reply vita3k/interface.cpp]: ah, drop, forget this smart idea, is not possible, obligatory need use read app file and read file separed sorry :/  here is impossible read module inside game lol

Comment (sunho) [reply vita3k/interface.cpp:301]:
```
bool is_app = fs::exists(module_app_path) && !fs::is_empty(module_app_path);
 if (is_app) {
   prefix = std::string("app0:");
        const char *const lib_load_list[] = {
            "sce_module/libc.suprx",
            "sce_module/libfios2.suprx",
            "sce_module/libult.suprx",
        };

} else {
  prefix = std::string("vs0:");
 const char *const lib_load_list[] = {
            "sys/external/libc.suprx",
            "sys/external/libfios2.suprx",
            "sys/external/libult.suprx",
        };
}

        for (auto &module_path : lib_load_list) {
            vfs::FileBuffer module_buffer;
            Ptr<const void> lib_entry_point;
            bool error;
             if (is_app)
       error = vfs::read_app_file(module_buffer, host.pref_path, host.io.title_id, module_path);
            else
  error = vfs::read_file(VitaIoDevice::vs0, module_buffer, host.pref_path, host.io.title_id, module_path);
        
            if (error) {
                SceUID module_id = load_self(lib_entry_point, host.kernel, host.mem, module_buffer.data(), prefix + module_path, host.cfg);
                if (module_id >= 0) {
                    const auto module = host.kernel.loaded_modules[module_id];
                    LOG_INFO("Pre-load module {} (at \"{}\") loaded", module->module_name, module_path);
                } else
                    return FileNotFound;
            } else {
                LOG_DEBUG("Pre-load module at \"{}\" not present", module_path);
            }
        }
```

Comment (Zangetsu38) [reply vita3k/interface.cpp:301]: no very like this view, drafted i see later

File: vita3k/modules/SceRtc/SceRtcUser.cpp [hunk 3]
```diff
@@ -118,55 +97,27 @@ EXPORT(int, sceRtcFormatRFC3339LocalTime) {
...
 
 EXPORT(int, sceRtcGetCurrentTick, SceRtcTick *tick) {
-    if (tick == nullptr) {
-        return RET_ERROR(SCE_RTC_ERROR_INVALID_POINTER);
-    }
-
-    tick->tick = rtc_get_ticks(host.kernel.base_tick.tick);
-
-    return 0;
+    return export_sceRtcGetCurrentTick(host, thread_id, export_name, tick);
```
Comment (frangarcj) [inline vita3k/modules/SceRtc/SceRtcUser.cpp:120]: Use the CALL_EXPORT define. this is failling cause you are missing a _ in the name. 
Also export_name is not correct.

File: vita3k/modules/SceLibKernel/SceLibKernel.h [hunk 4]
```diff
@@ -19,6 +19,8 @@
 
 #include <module/module.h>
 
+#include <../SceKernelThreadMgr/SceThreadmgr.h>
```
Comment (Zangetsu38) [inline vita3k/modules/SceLibKernel/SceLibKernel.h:22]: i have thinks used thyis, but also thinks is hack lol

**Comment by frangarcj** [comment]
This PR needs to move the encode_arm_inst to a common place with load_self.

If someone could care about that...

**Review by frangarcj — APPROVED** [review]

**Pull request merged**